samtranslator/model/eventsources/push.py

```diff
@@ -921,7 +924,29 @@ def _add_swagger_integration(  # type: ignore[no-untyped-def] # noqa: too-many-s
                 path=self.Path, method_name=self.Method, request_parameters=parameters
             )
 
-        api["DefinitionBody"] = editor.swagger
+        if api.get("MergeDefinitions", False):
+            self._merge_definitions(api, editor.swagger)
+        else:
+            api["DefinitionBody"] = editor.swagger
+
+    def _merge_definitions(self, api: Dict[str, Any], generated_body: Dict[str, Any]) -> None:
+        """
+        Merge SAM generated swagger definition into inline DefinitionBody using the follow rule:
+        - for a conflicting key, use SAM generated value
+        - otherwise include key-value pairs from both definitions
+        """
+        paths_property = api["DefinitionBody"]["paths"]
+
+        if self.Path not in paths_property:
+            paths_property.setdefault(self.Path, {})
```

Comment:
I have a hard time getting an intuitive sense of what it's doing, because it mutates the inputs, and uses instance attributes (so I need to know what `self.Path`, etc. are). Can it be made pure (no mutations) instead?